build.sh

```diff
@@ -3,28 +3,100 @@
 # Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
 #-------------------------------------------------------------------------------------------------------
 
-if [ ! -d "BuildLinux" ]; then
-    mkdir BuildLinux;
-fi
+SAFE_RUN() {
+    local SF_RETURN_VALUE=$($1 2>&1)
 
-pushd BuildLinux > /dev/null
+    if [[ $? != 0 ]]; then
+        >&2 echo $SF_RETURN_VALUE
+        exit
+    fi
+    echo $SF_RETURN_VALUE
+}
+
+PRINT_USAGE() {
+    echo ""
+    echo "build.sh [options]"
+    echo ""
+    echo "options:"
+    echo "--help, -h : Show help"
+    echo "--debug, -d : Debug build"
+    echo "CXX=<path to clang++>"
+    echo "CC=<path to clang>"
+    echo ""
+    echo "example: ./build.sh CXX=/path/to/clang++ CC=/path/to/clang"
+}
 
+_CXX=""
+_CC=""
 DEBUG_BUILD=0
-while getopts ":d" opt; do
-    case $opt in
-        d)
+
+while test $# -gt 0; do
+    if [[ "$1" =~ "CXX=" ]]; then
+        _CXX=$1
+        _CXX=${_CXX:4}
+    fi
+
+    if [[ "$1" =~ "CC=" ]]; then
+        _CC=$1
+        _CC=${_CC:3}
+    fi
+
+    if [[ "$1" == "--help" || "$1" == "-h" ]]; then
+        PRINT_USAGE
+        exit
+    fi
+
+    if [[ "$1" == "--debug" || "$1" == "-d" ]]; then
         DEBUG_BUILD=1
-        ;;
-    esac
+    fi
+
+    shift
 done
 
+if [[ ${#_CXX} > 0 && ${#_CC} > 0 ]]; then
+    echo "Custom CXX ${_CXX}"
+    echo "Custom CC  ${_CC}"
+
+    if [[ ! -f $_CXX || ! -f $_CC ]]; then
+        echo "ERROR: Custom compiler not found on given path"
+        exit 1
+    fi
+else
+    RET_VAL=$(SAFE_RUN 'c++ --version')
+    if [[ ! $RET_VAL =~ "clang" ]]; then
+        echo "Searching for Clang..."
+        if [[ ! -f /usr/bin/clang++ ]]; then
+            echo "ERROR: clang++ not found at /usr/bin/clang++"
+            echo ""
+            echo "You could use clang++ from a custom location."
+            PRINT_USAGE
+            exit 1
+        else
+            echo "Clang++ found at /usr/bin/clang++"
+            _CXX=/usr/bin/clang++
+            _CC=/usr/bin/clang
+        fi
```

Comment:
1. Positive if statement conditions are easier for the reader to parse, and there are both blocks here.
2. The success case is the goal, after all, makes sense for it to be the first thing to read and handle error as an alternative.
3. Personal preference in this case based on the above.